massa-execution/src/worker.rs

```diff
@@ -255,106 +253,137 @@ impl ExecutionWorker {
+                        // check whether there is a CSS-final block later in the same thread
+                        let mut check_slot = Slot::new(cur_slot.period + 1, cur_slot.thread);
+                        while check_slot <= max_css_final_slot {
+                            if self.pending_css_final_blocks.contains_key(&check_slot) {
+                                break;
+                            }
+                            check_slot.period += 1;
+                        }
+                        if check_slot <= max_css_final_slot {
+                            // subsequent CSS-final block found in the same thread as cur_slot
+                            // execute a miss as an SCE-final VM step
+                            self.push_request(ExecutionRequest::RunFinalStep(ExecutionStep {
+                                slot: cur_slot,
+                                block: None,
+                            }));
+                            // update cursors
+                            self.last_active_slot = cur_slot;
+                            self.last_final_slot = cur_slot;
+                        } else {
+                            // no subsequent CSS-final block found in the same thread as cur_slot
+                            break;
+                        }
+                    }
 
-        // list SCE-final slots/blocks
-        for (b_id, block) in css_final_blocks.into_iter() {
-            let block_slot = block.header.content.slot;
-            if block_slot <= self.last_final_slot {
-                continue;
-            }
-            loop {
-                let next_final_slot = self.last_final_slot.get_next_slot(self.cfg.thread_count)?;
-                if next_final_slot == block_slot {
-                    self.push_request(ExecutionRequest::RunFinalStep(ExecutionStep {
-                        slot: next_final_slot,
-                        block: Some((b_id, block.clone())),
-                    }));
-                    self.last_final_slot = next_final_slot;
-                    break;
-                } else if next_final_slot < max_thread_slot[next_final_slot.thread as usize] {
-                    self.push_request(ExecutionRequest::RunFinalStep(ExecutionStep {
-                        slot: next_final_slot,
-                        block: Some((b_id, block.clone())),
-                    }));
-                    self.last_final_slot = next_final_slot;
-                } else {
-                    self.ordered_pending_css_final_blocks.push((b_id, block));
-                    break;
+                    // there are no more CSS-final blocks
+                    None => break,
                 }
+
+                cur_slot = cur_slot.get_next_slot(self.cfg.thread_count)?;
             }
         }
 
-        // list remaining CSS finals + new blockclique
-        self.ordered_active_blocks = self
-            .ordered_pending_css_final_blocks
+        // process SCE-active blocks
```

Comment:
This, or the spec, should be updated so that they match, I believe the spec currently reads `blocks that remain to be processed are in blockclique_blocks UNION pending_css_final_blocks`

Reply:
done